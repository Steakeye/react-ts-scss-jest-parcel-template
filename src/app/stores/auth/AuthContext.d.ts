import React from 'react';
import '@auth/auth-spa-js';
import { OptionalN } from '~/app/declarations/standard';

interface AuthData {
  isReady: boolean;
  isAuthenticated: boolean;
  isAuthError: boolean;
  user: unknown;
  loading: boolean;
}

export interface AuthRedirectAppState<S = unknown> {
  originURL: string;
  state?: S;
}

interface AuthInternalData extends AuthData {
  appState?: AuthRedirectAppState;
}

interface AuthContextValue extends AuthData {
  //TODO
}

type AuthContext = React.Context<OptionalN<AuthContextValue>>;

type AuthHook = () => OptionalN<AuthContextValue>;
