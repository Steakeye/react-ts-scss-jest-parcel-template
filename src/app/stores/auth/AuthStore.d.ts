import * as React from 'react';
import { RouteComponentProps } from "react-router";
import { ImmutableDataReducerActionTypes } from "~/app/global/models.d";
import { AuthData, AuthInternalData, AuthRedirectAppState } from "~/app/stores/auth/AuthContext.d";

//Temp
interface TempAuthClientOptions {}

type AuthReducerDispatchActionType = ImmutableDataReducerActionTypes | 'ready';

interface AuthReducerDispatchAction extends Partial<AuthInternalData> {
  type: AuthReducerDispatchActionType;
}

interface AuthProviderProps extends Partial<TempAuthClientOptions> {
  children: React.ReactNode;
  onRedirectCallback?: DefaultRedirectCallback;
}

type AuthStoreProps = React.PropsWithChildren<AuthProviderProps>

type AuthStoreFC = (
  props: AuthStoreProps & RouteComponentProps
) => JSX.Element;

type DefaultRedirectCallback = (result: AuthRedirectAppState) => void;
