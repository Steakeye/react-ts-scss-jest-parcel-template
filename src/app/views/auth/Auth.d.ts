import { RouteComponentProps } from "react-router";

interface AuthFCProps {
  userID?: string;
  tags?: string;
}

export type AuthFC = (props: RouteComponentProps<AuthFCProps>) => JSX.Element;
