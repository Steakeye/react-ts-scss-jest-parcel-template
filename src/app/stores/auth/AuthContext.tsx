import React, { useContext } from 'react';
import {
  AuthContextValue,
  AuthHook,
  AuthContext,
} from '~/app/stores/auth/AuthContext.d';
import { OptionalN } from '~/app/declarations/standard';

export const authContext: AuthContext = React.createContext<
  OptionalN<AuthContextValue>
>(null);

export const useAuth: AuthHook = () =>
  useContext<OptionalN<AuthContextValue>>(authContext);

export { authContext as AuthContext };
