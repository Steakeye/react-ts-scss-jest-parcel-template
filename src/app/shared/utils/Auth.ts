import { getApplicationRootURL } from '~/app/shared/utils/AppDOM';
import { AuthRedirectAppState } from '~/app/stores/auth/AuthContext.d';
import { AUTH_ROUTE } from '~/app/AppRoutes';

export interface RedirectLoginOptions {
  redirect_uri: string;
  appState: AuthRedirectAppState;
}

export function redirectToAuth(
  location: Location,
  authRedirect: (options?: RedirectLoginOptions) => void
) {
  const root = getApplicationRootURL();
  const { pathname, search, hash } = location;

  const appState: AuthRedirectAppState = {
    originURL: `${pathname}${search}${hash}`,
  };

  authRedirect({
    redirect_uri: `${root}${AUTH_ROUTE}`,
    appState,
  });
}
