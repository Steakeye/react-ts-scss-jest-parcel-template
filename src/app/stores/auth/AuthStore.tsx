import React, { useEffect, useReducer } from 'react';
//import { Config } from '~/app/Config';
import { AuthContext } from './AuthContext';
import {
  AuthReducerDispatchAction,
  AuthStoreProps,
  AuthStoreFC,
} from './AuthStore.d';
import { AuthInternalData } from './AuthContext.d';
import { RouteComponentProps, withRouter } from 'react-router';

//Temp
//interface TempAuthClient {}

//const { auth: authConfig } = Config;

function createInitialState(): AuthInternalData {
  return {
    isReady: false,
    isAuthenticated: false,
    isAuthError: false,
    user: undefined,
    loading: true,
    popupOpen: false,
  };
}

function authReducer(
  state: AuthInternalData,
  action: AuthReducerDispatchAction
): AuthInternalData {
  let updatedState: AuthInternalData;

  //TODO: Reducer - implementation
  updatedState = state;

  return updatedState;
}

export const AuthStore = ({
  children,
  history,
  ...initOptions
}: AuthStoreProps & RouteComponentProps) => {
  //const [authClient, setAuthClient] = useState<TempAuthClient>(); //TODO
  const [data /*setData*/] = useReducer(authReducer, createInitialState());

  useEffect(() => {
    (async () => {
      //TODO: Setup - implementation
    })();
  }, []);

  useEffect(() => {
    //TODO: Update - implementation
  }, [data]);

  const { isReady, isAuthenticated, isAuthError, user, loading } = data;

  return (
    <AuthContext.Provider
      value={{
        isReady,
        isAuthenticated,
        isAuthError,
        user,
        loading,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};

const authStoreWithRouter = withRouter<
  AuthStoreProps & RouteComponentProps,
  AuthStoreFC
>(AuthStore);

export { authStoreWithRouter as AuthStoreWithRouter };
