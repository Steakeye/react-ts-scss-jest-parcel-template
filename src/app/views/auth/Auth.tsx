import * as React from 'react';
// eslint-disable-next-line node/no-extraneous-import
import {Navigate,} from 'react-router';
import {useSearchParams } from 'react-router-dom';
import { Helmet } from 'react-helmet';
import { Views as ViewStrings } from '~/app/AppStrings';
import { ERROR_ROUTE } from '~/app/AppRoutes';
import { AuthContextValue } from '~/app/stores/auth/AuthContext.d';
import { useAuth } from '~/app/stores/auth/AuthContext';

//Temp
const REDIRECT_QUERY_KEY = 'TEMP_REDIRECT_QUERY_KEY';

const { Auth: authViewStrings } = ViewStrings;

const { MainHeading: mainHeading } = authViewStrings;

const pageTitle = `AppName - ${mainHeading}`;

function authMetaDataComponents(): JSX.Element {
  return (
    <>
      <Helmet>
        <title>{pageTitle}</title>
      </Helmet>
    </>
  );
}

function Auth() {
  const query = useSearchParams()[0]
  const redirectPath = query.get(REDIRECT_QUERY_KEY);

  const { isAuthError } = useAuth() as AuthContextValue;

  let view: JSX.Element;

  if (isAuthError) {
    view = <Navigate to={ERROR_ROUTE} replace />;
  } else if (redirectPath) {
    view = <Navigate to={redirectPath as string} replace />;
  } else {
    view = (
      <>
        {/* Meta content */}
        {authMetaDataComponents()}
        {/* Body content */}
        <main></main>
      </>
    );
  }

  return view;
}

export { Auth };
