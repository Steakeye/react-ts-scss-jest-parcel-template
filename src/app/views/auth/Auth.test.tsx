import * as React from 'react';
// eslint-disable-next-line node/no-extraneous-import
import { Navigate, NavigateProps } from 'react-router';
import { render, RenderResult } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import {
  createRouteComponentProps,
  createMockComponent,
} from '~/app/../../jest-utils/Helpers';
import { useAuth } from '~/app/stores/auth/AuthContext';
import { AuthFCProps, AuthFC } from './Auth.d';

jest.mock('~/app/stores/auth/AuthContext', () => ({
  useAuth: jest.fn().mockReturnValue({ isAuthError: false }),
}));

jest.mock('~/app/AppRoutes', () => ({
  REDIRECT_QUERY_KEY: 'mockRedirectKey',
  ERROR_ROUTE: '/mockErrorPath',
}));

jest.mock('~/app/AppStrings', () => ({
  Views: {
    Auth: {
      MainHeading: 'mock_main_heading',
    },
  },
}));

const queryGetter = jest.fn();

jest.doMock('react-router', () => ({
  Navigate: createMockComponent('Navigate', 'mock-navigate'),
}));

jest.doMock('react-router-dom', () => ({
  useSearchParams: jest.fn(() => [{ get: queryGetter }]),
}));

describe('Auth view - handles auth', () => {
  let Auth: AuthFC;
  let authInstance: RenderResult;
  let navigate: (props: NavigateProps) => typeof Navigate;

  beforeAll(async () => {
    ({ Auth } = await import('./Auth'));
    navigate = ((await import('react-router')).Navigate as unknown) as (
      props: NavigateProps
    ) => typeof Navigate;
  });

  describe('default behaviour', () => {
    beforeAll(() => {
      const defaultProps = createRouteComponentProps<AuthFCProps>();
      authInstance = render(<Auth {...defaultProps} />);
    });

    afterAll(() => {
      (queryGetter as jest.Mock).mockClear();
    });

    it('renders without error', () => {
      const el = authInstance.container;
      expect(el).toMatchSnapshot();
    });

    it('the auth hook is used', () => {
      expect(useAuth).toHaveBeenCalled();
      expect(useAuth).toHaveBeenCalledTimes(1);
      expect(useAuth).toHaveBeenCalledWith();
    });

    it('the query string is parsed', () => {
      expect(queryGetter).toHaveBeenCalled();
      expect(queryGetter).toHaveBeenCalledTimes(1);
      expect(queryGetter).toHaveBeenCalledWith('TEMP_REDIRECT_QUERY_KEY');
    });
  });

  describe('when a navigate value is passed', () => {
    beforeAll(() => {
      (queryGetter as jest.Mock).mockReturnValue('/mockPath');
      authInstance = render(<Auth />);
    });

    it('renders without error', () => {
      const el = authInstance.container;
      expect(el).toMatchSnapshot();
    });

    it('the query string is parsed', () => {
      expect(queryGetter).toHaveBeenCalled();
      expect(queryGetter).toHaveBeenCalledTimes(1);
      expect(queryGetter).toHaveBeenCalledWith('TEMP_REDIRECT_QUERY_KEY');
    });

    it('and the navigate component is used', () => {
      expect(navigate).toHaveBeenCalled();
      expect(navigate).toHaveBeenCalledTimes(1);
      expect(navigate).toHaveBeenCalledWith(
        { replace: true, to: '/mockPath' },
        {}
      );
    });
  });

  describe('when an auth error occurs', () => {
    beforeAll(() => {
      (useAuth as jest.Mock).mockReturnValue({
        isAuthError: true,
      });
      (queryGetter as jest.Mock).mockClear();
      (navigate as jest.Mock).mockClear();

      authInstance = render(<Auth />);
    });

    it('renders without error', () => {
      const el = authInstance.container;
      expect(el).toMatchSnapshot();
    });

    it('the query string is parsed', () => {
      expect(queryGetter).toHaveBeenCalled();
      expect(queryGetter).toHaveBeenCalledTimes(1);
      expect(queryGetter).toHaveBeenCalledWith('TEMP_REDIRECT_QUERY_KEY');
    });

    it('and the navigate component is used', () => {
      expect(navigate).toHaveBeenCalled();
      expect(navigate).toHaveBeenCalledTimes(1);
      expect(navigate).toHaveBeenCalledWith(
        { replace: true, to: '/mockErrorPath' },
        {}
      );
    });
  });
});
