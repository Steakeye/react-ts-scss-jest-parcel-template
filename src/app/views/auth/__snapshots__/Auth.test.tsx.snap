// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth view - handles auth default behaviour renders without error 1`] = `
<div>
  <main />
</div>
`;

exports[`Auth view - handles auth when a redirect value is passed renders without error 1`] = `
<div>
  <span
    data-mock="Redirect"
    data-testid="mock-redirect"
    to="/mockPath"
  >
    Mock Redirect
  </span>
</div>
`;

exports[`Auth view - handles auth when an auth error occurs renders without error 1`] = `
<div>
  <span
    data-mock="Redirect"
    data-testid="mock-redirect"
    to="/mockErrorPath"
  >
    Mock Redirect
  </span>
</div>
`;
